File: app/src/main/java/com/jbrunton/mymovies/api/DescriptiveError.java
package com.jbrunton.mymovies.api;

public class DescriptiveError extends RuntimeException {
    private final boolean isNetworkError;

    public DescriptiveError(String message, boolean isNetworkError) {
        super(message);
        this.isNetworkError = isNetworkError;
    }

    public DescriptiveError(String message, Throwable cause, boolean isNetworkError) {
        super(message, cause);
        this.isNetworkError = isNetworkError;
    }

    public boolean isNetworkError() {
        return isNetworkError;
    }

    public static DescriptiveError from(Throwable throwable) {
        if (throwable instanceof DescriptiveError) {
            return (DescriptiveError) throwable;
        } else {
            return new DescriptiveError(
                    messageFromThrowable(throwable),
                    throwable,
                    false);
        }
    }

    private static String messageFromThrowable(Throwable throwable) {
        if (throwable.getMessage() == null) {
            return "Unexpected Error";
        }
        return throwable.getMessage();
    }
}

File: networking/src/main/java/com/jbrunton/networking/DescriptiveError.kt
package com.jbrunton.networking

class DescriptiveError : RuntimeException {
    val isNetworkError: Boolean

    constructor(message: String, isNetworkError: Boolean) : super(message) {
        this.isNetworkError = isNetworkError
    }

    constructor(message: String, cause: Throwable, isNetworkError: Boolean) : super(message, cause) {
        this.isNetworkError = isNetworkError
    }

    companion object {

        fun from(throwable: Throwable): DescriptiveError {
            return throwable as? DescriptiveError ?: DescriptiveError(
                    messageFromThrowable(throwable),
                    throwable,
                    false)
        }

        private fun messageFromThrowable(throwable: Throwable) = throwable.message ?: "Unexpected Error"
    }

    override val message: String
        get() = super.message!!
}